jupyterhub/handlers/base.py

```diff
@@ -1348,7 +1349,13 @@ def template_namespace(self):
             xsrf=self.xsrf_token.decode('ascii'),
         )
         if self.settings['template_vars']:
-            ns.update(self.settings['template_vars'])
+            for key, value in self.settings['template_vars'].items():
+                if callable(value):
```

Comment:
So far, all callable hooks have a single defined signature, rather than variable by inspection, which simplifies things a bit. What about calling it with `f(user)` unconditionally?

And should they be allowed to be async? If so, add:

```python
            if inspect.isawaitable(value):
                value = await value
```

Reply:
ah, I recalled that templates may be async, but for backward-compatibility, `render_template` must be _allowed_ to be called synchronously. I'll think about that. Let's go with sync-only for now.

I think if you return awaitables from the template namespace, you'll be able to await them _in_ the template, assuming the page is rendered async (all default pages are, I believe).